### DISCUSSION TITLE
How Yarn v2 Workspace can resolve dependencies from registry gitlab

### DISCUSSION THREAD

**Discussion by Piero87** [yarnpkg/berry#3063] [Q&A]
I'm trying to create a monorepo with yarn v2 workspace, and then I want publish each packages inside gitlab packages registry to avoid build each time a packages if there is already the same version in the registry. My project example is public on gitlab: https://gitlab.com/p.nompleggio/monorepov3

With this configuration all works fine, these two lines in `.gitlab-ci.yml`:

```
yarn workspaces foreach -pt run build
yarn workspaces foreach --no-private npm publish
```

All packages are build and published inside the registry so it's all right.

But now I'm trying to build only changed package inside the workspace and not only all packages. So I've tried just for understanding how all works to mantain this package inside registry `@monorepov3/common@1.0.1` and edit the .gitlab-ci.yml in this way:

```
yarn workspace @monorepov3/backoffice-app run build
yarn workspace @monorepov3/backoffice-app npm publish
```

So I have tried to build only the` @monorepov3/backoffice-app` package and I wanna retrieve `@monorepov3/common` dependices inside `@monorepov3/backoffice-app` from the registry, but when the build command start doesn't work because doesn't find the `@monorepov3/common` locally and doesn't look in the registry.

Can anyone tell me how can I achieve this? How can I build with the package inside the registry for dependencies inside the ci/cd pipeline?

Another question is how can I build only the packages that are changed?

**Comment by merceyz** [comment]
> How can I build with the package inside the registry for dependencies inside the ci/cd pipeline?

You can set `enableTransparentWorkspaces: false` on the CI to make it fetch from the registry when the `workspace:` protocol isn't explicitly specified
https://yarnpkg.com/configuration/yarnrc#enableTransparentWorkspaces

> how can I build only the packages that are changed?

See https://github.com/yarnpkg/berry/issues/2374

**Reply by Piero87** [reply]
@merceyz Thank you for the answer, but doesn't work, I don't know If i'm doing something wrong now I'm trying locally (that would be the same then in a CI), this is the package.json of `@monorepov3/backoffice-app`:

  "dependencies": {
    "@monorepov3/common": "workspace:packages/common",
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  }

I have set in .yarnrc.yml this:

`enableTransparentWorkspaces: false`

I delete locally the dist folder of the build of `@monorepov3/common`, and when I build the `@monorepov3/backoffice-app` with this command:

`yarn workspace @monorepov3/backoffice-app run build`

the build fail because doesn't find the build of `@monorepov3/common` (I've tried also in CI and I have the same error).

In the registry (if you wanna see you can check with the link above is public) there is the pack of `@monorepov3/common` @ 1.0.1

If I change this:
```
  "dependencies": {
    "@monorepov3/common": "workspace:packages/common",
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  }
```

to this:
```
  "dependencies": {
    "@monorepov3/common": "1.0.1",
    "react": "^17.0.2",
    "react-dom": "^17.0.2"
  }
```

and I launch a yarn install it will retrieve it from the registry, but with this  `"workspace:packages/common"` doesn't, there something I am missing with the configuration? 
How can this: `"workspace:packages/common"` become this `"1.0.1"` or the latest version found in the registry?

**Reply by merceyz** [reply]
> and I launch a yarn install it will retrieve it from the registry, but with this "workspace:packages/common" doesn't, there something I am missing with the configuration?

That's correct and how it's supposed to behave

> How can this: "workspace:packages/common" become this "1.0.1" or the latest version found in the registry?

You can't, that will always resolve to the local version, based on that you should be able to detect changed workspaces somehow and then change the declared dependencies before you install. There is nothing built-in to do this for you so you'll need to write some scripts
